R/object_name_linters.R

```diff
@@ -44,15 +44,36 @@ object_name_xpath <- local({
 #' @param styles A subset of
 #'   \Sexpr[stage=render, results=rd]{lintr:::regexes_rd}. A name should
 #'   match at least one of these styles.
+#' @param regexes A (possibly named) character vector specifying a custom naming convention.
+#'   If named, the names will be used in the lint message. Otherwise, "custom" will be used as a name for the style.
+#'   Quotes (`` `"' ``) and specials (`%` and trailing `<-`) are not considered part of the name to be matched.
 #' @evalRd rd_tags("object_name_linter")
 #' @seealso [linters] for a complete list of linters available in lintr.
 #' @export
-object_name_linter <- function(styles = c("snake_case", "symbols")) {
-  styles <- match.arg(styles, names(style_regexes), several.ok = TRUE)
+object_name_linter <- function(styles = c("snake_case", "symbols"), regexes = character()) {
+  if (length(styles) > 0L) {
+    # Allow `object_name_linter(NULL, "my_regex")`
+    styles <- match.arg(styles, names(style_regexes), several.ok = TRUE)
+    style_list <- style_regexes[styles]
+  } else {
+    style_list <- list()
+  }
+  if (length(regexes) > 0L) {
+    if (!is.character(regexes)) {
+      stop("`regexes` must be a character vector.")
+    }
+    if (is.null(names(regexes))) {
+      names(regexes) <- "custom"
+    }
+    style_list <- c(style_list, as.list(regexes))
```

Comment:
what happens if user provides an exact replica of one of the existing `style_regexes`? do we need `unique()`?

Reply:
Not really. There is no harm with a name matching multiple styles.

Reply:
What does the message look like for matching multiple "custom" styles?

Reply:
See L76, duplicates are removed in the message, so the message will read "Variable and function name style should be custom."

Reply:
Got it. Would still like to see some tests around this to ensure that's what's happening / continues to happen.